monitoring/ping.go

```diff
@@ -0,0 +1,307 @@
+	latencyStatsSlidingWindowSize = 20
+	// pingMinimum sets the minimum value that can be recorded
+	pingMinimum = 0 * time.Second
+	// pingMaximum sets the maximum value that can be recorded
+	pingMaximum = 10 * time.Second
+	// pingSignificantFigures specifies how many decimals should be recorded
+	pingSignificantFigures = 3
+	// latencyThreshold sets the RTT threshold
+	latencyThreshold = 15 * time.Millisecond
+	// latencyQuantile sets the quantile used while checking Histograms against RTT results
+	latencyQuantile = 95.0
+)
+
+// pingChecker is a checker that verifies that ping times (RTT) between nodes in
+// the cluster are within a predefined threshold
+type pingChecker struct {
+	self           serf.Member
+	serfClient     agent.SerfClient
+	serfRPCAddr    string
+	serfMemberName string
+	latencyStats   ttlmap.TTLMap
+	mux            sync.Mutex
+	logger         log.Entry
+}
+
+// PingCheckerConfig is used to store all the configuration related to the current check
+type PingCheckerConfig struct {
+	// SerfRPCAddr is the address used by the Serf RPC client to communicate
+	// with the Serf cluster
+	SerfRPCAddr string
+	// SerfMemberName is the name assigned to this node in Serf
+	SerfMemberName string
+	// NewSerfClient is an optional Serf Client function that can be used instead
+	// of the default one. If not specified it will fallback to the default one
+	NewSerfClient agent.NewSerfClientFunc
+}
+
+// CheckAndSetDefaults validates that this configuration is correct and sets
+// value defaults where necessary
+func (c *PingCheckerConfig) CheckAndSetDefaults() error {
+	if c.SerfRPCAddr == "" {
+		return trace.BadParameter("serf rpc address can't be empty")
+	}
+	if c.SerfMemberName == "" {
+		return trace.BadParameter("serf member name can't be empty")
+	}
+	if c.NewSerfClient == nil {
+		c.NewSerfClient = agent.NewSerfClient
+	}
+	return nil
+}
+
+// NewPingChecker returns a checker that verifies accessibility of nodes in the cluster by exchanging ping requests
+func NewPingChecker(conf PingCheckerConfig) (c health.Checker, err error) {
+	err = conf.CheckAndSetDefaults()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	latencyTTLMap, err := ttlmap.New(latencyStatsCapacity)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	logger := log.WithFields(log.Fields{trace.Component: "ping"})
+	logger.Debugf("using Serf IP: %v", conf.SerfRPCAddr)
+	logger.Debugf("using Serf Name: %v", conf.SerfMemberName)
+
+	client, err := conf.NewSerfClient(serf.Config{
+		Addr: conf.SerfRPCAddr,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// retrieve other nodes using Serf members
+	nodes, err := client.Members()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// finding what is the current node
+	var self serf.Member
+	for _, node := range nodes {
+		logger.Debugf("node %s status %s", node.Name, node.Status)
+		if node.Status != pb.MemberStatus_Alive.String() {
+			continue
+		}
```

Comment:
I don't think this is possible at this early stage - i.e. when the checkers are first added to the agent. I think at this point the agent is not yet part of the serf cluster.